lib/trie/trie.go

```diff
@@ -65,6 +102,46 @@ func (t *Trie) DeepCopy() (*Trie, error) {
 	return cp, nil
 }
 
+func isSimilar(tNode node, trieNode node) bool {
+	tBranch, ok := tNode.(*branch)
+	trieBranch, trieOk := trieNode.(*branch)
+	if ok && trieOk {
+		if tBranch.value != nil && trieBranch.value != nil {
+			if !bytes.Equal(tBranch.value, trieBranch.value) {
+				return false
+			}
+		}
+		for tKey, child := range tBranch.children {
+			if child != nil && trieBranch.children[tKey] != nil {
+				if !isSimilar(child, trieBranch.children[tKey]) {
+					return false
+				}
+				continue
+			} else if child == nil && trieBranch.children[tKey] == nil {
+				continue
+			}
+			return false
+		}
+		return true
+	}
+	tLeaf, ok := tNode.(*leaf)
+	trieLeaf, trieOk := trieNode.(*leaf)
+	if ok && trieOk {
+		if tLeaf.value == nil && trieLeaf.value == nil {
+			return true
+		} else if tLeaf.value != nil && trieLeaf.value != nil {
+			return bytes.Equal(tLeaf.value, trieLeaf.value)
+		}
+		return false
+	}
+	return false
+}
+
+// Equal returns true if both tries are same.
+func (t *Trie) Equal(trie *Trie) bool {
+	return isSimilar(t.root, trie.root)
```

Comment:
why is this needed? can we not just compare the root hashes?

Reply:
Silly me 😅 It skipped my mind that the trie is also a Merkle tree.